### PR TITLE
feat: Upload file & Testing

### LINKED ISSUE #10: feat: upload file
## Tasks

- [ ] Create the struct in [RMI](https://github.com/hawks-atlanta/rmi-java) repository
- [ ] Update the gateway [dev](https://github.com/hawks-atlanta/gateway-java/tree/dev) branch
- [ ] Implement the service [worker](https://github.com/hawks-atlanta/worker-java)

## Ideas

```
struct upload {
     string uuid;
     byte[] contents
}
```

### PR DESCRIPTION
**Pull request by Woynert**
## Includes:

- Implement write file methods
- Related tests added
	- (Test with direct RMI client pending)
- Added docker-compose for testing
- Added new env var `METADATA_BASEPATH`

### :open_file_folder: Setup directory structure

1. Setup a base directory for the files to be saved. It must have the following structure:

```
.
├── backups
│   ├── volume1
│   ├── volume2
│   └── volume3
└── files
    ├── volume1
    ├── volume2
    └── volume3
```

You can create this structure with the provided script `.setup-vol-dir` it receives a `path` to create the structure in. Usage:

```sh
$ ./.setup-vol-dir ./capyfile
$ tree capyfile/
capyfile/
├── backups
│   ├── volume1
│   ├── volume2
│   └── volume3
└── files
    ├── volume1
    ├── volume2
    └── volume3
```

### :package: Starting services

```sh
docker-compose up
```

### :test_tube: Running tests

For running the Worker, it needs the following two variables set:

- `AVAILABLE_VOLUMES`: Which volumes are available
- `VOLUME_BASE_PATH`: Full path to volume base path

```sh
AVAILABLE_VOLUMES=1,2,3  VOLUME_BASE_PATH=/misc/worker-java/capyfile  ./gradlew test
```

After that a new file should have been created:

```
$ tree capyfile/
capyfile/
├── backups
│   ├── volume1
│   ├── volume2
│   │   └── 209ad127-3eb4-4217-9e00-2b0454f676b6
│   └── volume3
└── files
    ├── volume1
    ├── volume2
    │   └── 209ad127-3eb4-4217-9e00-2b0454f676b6
    └── volume3
```

### CODE REVIEW

File: app/src/main/java/worker/config/Config.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		String aVols = getEnv ("AVAILABLE_VOLUMES", "");
+
+		try {
+			Config.availableVolumes =
+				Arrays.stream (aVols.split (",")).mapToInt (Integer::parseInt).toArray ();
+			if (availableVolumes.length == 0) {
+				throw new RuntimeException ("No volumes available");
+			}
+		} catch (Exception e) {
+			throw new RuntimeException ("No volumes available");
```
Comment (PedroChaparro) [inline app/src/main/java/worker/config/Config.java]: I think having a different error would be better to ease the debug of non-valid `AVAILABLE_VOLUMES` strings

File: app/src/test/java/worker/UTFileIO.java
Comment (PedroChaparro) [inline app/src/test/java/worker/UTFileIO.java:1]: Could you add another test to simulate saving a directory? Just to ensure that also works during development

**Review by PedroChaparro — COMMENTED** [review]
Good work 👏🏽. It would be nice if you can make the following updates to the `README` file: 

- Add the `-x` flag in the build step to skip the testing tag (`./gradlew build -x test`)
- Add the `--rerun` option to re-run tests ignoring previous executions / cache (`./gradlew test --rerun`).

File: app/src/main/java/worker/config/Config.java (hunk 1 above)
Comment (Woynert) [reply app/src/main/java/worker/config/Config.java]: Do you mean a new type of error or a new message?

File: app/src/test/java/worker/UTFileIO.java
Comment (Woynert) [reply app/src/test/java/worker/UTFileIO.java:1]: The Worker service doesn't save directories, only files

File: app/src/main/java/worker/config/Config.java (hunk 1 above)
Comment (PedroChaparro) [reply app/src/main/java/worker/config/Config.java]: Just a different error message

File: app/src/test/java/worker/UTFileIO.java
Comment (PedroChaparro) [reply app/src/test/java/worker/UTFileIO.java:1]: True, I always forget that

**Review by PedroChaparro — APPROVED** [review]

**Pull request merged**